client/admin.go

```diff
@@ -0,0 +1,196 @@
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Marc Berhault (marc@cockroachlabs.com)
+
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/base"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+const (
+	// Types of admin config to manipulate.
+
+	// Accounting expects proto.AcctConfig
+	Accounting = "acct"
+	// Permission expects proto.PermConfig
+	Permission = "perms"
+	// Quit only handles Get requests.
+	Quit = "quit"
+	// User expects proto.UserConfig
+	User = "users"
+	// Zone expects proto.ZoneConfig.
+	Zone = "zones"
+)
+
+// AdminClient issues http requests to admin endpoints.
+// TODO(marc): unify the way we handle addresses in clients.
+type AdminClient struct {
+	context    *base.Context
+	address    string
+	configType string
+}
+
+// NewAdminClient returns a new 'configType' admin client, talking to 'address'.
+func NewAdminClient(ctx *base.Context, address, configType string) *AdminClient {
+	return &AdminClient{ctx, address, configType}
+}
+
+// adminURI build a URI using the passed-in key.
+func (a *AdminClient) adminURI(key string) string {
+	uri := fmt.Sprintf("%s://%s/_admin/%s/%s", a.context.RequestScheme(),
+		a.address, a.configType, url.QueryEscape(key))
+	return uri
+}
+
+// Get issues a GET and returns the plain-text body. It cannot take a key.
+func (a *AdminClient) Get() (string, error) {
+	// Construct our own URI: there is no terminating '/'.
+	uri := fmt.Sprintf("%s://%s/_admin/%s", a.context.RequestScheme(),
+		a.address, a.configType)
+	req, err := http.NewRequest("GET", uri, nil)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Add(util.AcceptHeader, util.PlaintextContentType)
+	body, err := a.do(req)
+	if err != nil {
+		return "", err
+	}
+	return string(body), nil
+}
+
+// GetJSON issues a GET request and returns a json-encoded response.
+func (a *AdminClient) GetJSON(key string) (string, error) {
+	req, err := http.NewRequest("GET", a.adminURI(key), nil)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Add(util.AcceptHeader, util.JSONContentType)
+	body, err := a.do(req)
+	if err != nil {
+		return "", err
+	}
+	return string(body), nil
+}
+
+// GetYAML issues a GET request and returns a yaml-encoded response.
+func (a *AdminClient) GetYAML(key string) (string, error) {
+	req, err := http.NewRequest("GET", a.adminURI(key), nil)
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Add(util.AcceptHeader, util.YAMLContentType)
+	log.Infof("***** Issuing: %+v", req)
+	body, err := a.do(req)
+	log.Infof("Got: %v, %v", body, err)
+	if err != nil {
+		return "", err
+	}
+	return string(body), nil
+}
+
+// SetYAML issues a POST request for the given key using the yaml-encoded body.
+func (a *AdminClient) SetYAML(key, body string) error {
```

Comment:
Ditto about explicit types. Proto will also be more interesting since I'll most like take/return a proto.message and do the marshaling myself.